File: DataStructures/SegTree.java
class SegTree {
    int[] seg, arr, lazy;
    public SegTree(int[] arr) {
        this.arr = arr;
        seg = new int[arr.length * 4];
        lazy = new int[arr.length * 4];
    }

    private int getSum(int l, int r, int a, int b, int node) {
        if (lazy[node] != 0) {
            seg[node] += (r - l + 1) * lazy[node];
            if (l != r) {
                lazy[node * 2] += lazy[node];
                lazy[node * 2 + 1] += lazy[node];
            }
            lazy[node] = 0;
        }

        if (l > r || l > b || r < a)
            return 0;
        if (l >= a && r <= b)
            return seg[node];

        int mid = (l + r) / 2;
        return getSum(l, mid, a, b, 2 * node) + getSum(mid + 1, r, a, b, 2 * node + 1);
    }

    int getSum(int l, int r) {
        return getSum(0, arr.length - 1, l, r, 1);
    }

    private void updateRange(int node, int l, int r, int a, int b, int diff) {
        if (lazy[node] != 0) {
            seg[node] += (r - l + 1) * lazy[node];
            if (l != r) {
                lazy[node * 2] += lazy[node];
                lazy[node * 2 + 1] += lazy[node];
            }
            lazy[node] = 0;
        }

        if (l > r || l > b || r < a)
            return;

        if (l >= a && r <= b) {
            seg[node] += (r - l + 1) * diff;
            if (l != r) {
                lazy[node * 2] += diff;
                lazy[node * 2 + 1] += diff;
            }
            return;
        }
        int mid = (l + r) / 2;
        updateRange(node * 2, l, mid, a, b, diff);
        updateRange(node * 2 + 1, mid + 1, r, a, b, diff);
        seg[node] = seg[node * 2] + seg[node * 2 + 1];
    }

    void updateRange(int a, int b, int diff)  {
        updateRange(1, 0, arr.length - 1, a, b, diff);
    }

    private void build(int l, int r, int node) {
        if (l > r)
            return;
        if (l == r) {
            seg[node] = arr[l];
            return;
        }
        int mid = (l + r) / 2;
        build(l, mid, node * 2);
        build(mid + 1, r, node * 2 + 1);

        seg[node] = seg[node * 2] + seg[node * 2 + 1];
    }
    void build() {
        build(0, arr.length - 1, 1);
    }
}

File: DataStructures_CPP/SegTree.cpp
class SegTree {
public:
    int len;
    vi seg, arr;

    SegTree(int l) {
        len = l;
        seg.resize(l * 4);
    }

    SegTree(vi a) {
        len = a.size();
        seg.resize(len * 4);
        arr = a;
        build(1, 0, len - 1);
    }

    int query(int a, int b) {
        return query(a, b, 1, 0, len - 1);
    }

    void update(int idx, int diff) {
        update(1, 0, len - 1, idx, diff);
    }

private:
    int f(int a, int b) { //change for different problems
        return a + b;
    }

    void build(int node, int l, int r) {
        if (l > r)
            return;
        if (l == r) {
            seg[node] = arr[l];
            return;
        }
        int mid = (l + r) / 2;
        build(node * 2, l, mid);
        build(node * 2 + 1, mid + 1, r);
        seg[node] = f(seg[node * 2], seg[node * 2 + 1]);
    }

    int query(int a, int b, int node, int l, int r) {
        if (l > b || r < a) {
            return -1;
        }
        if (l >= a && r <= b) {
            return seg[node];
        }
        int mid = (l + r) / 2;
        int n1 = query(a, b, node * 2, l, mid);
        int n2 = query(a, b, node * 2 + 1, mid + 1, r);
        if (n1 == -1)
            return n2;
        if (n2 == -1)
            return n1;
        return f(n1, n2);
    }

    void update(int node, int l, int r, int idx, int diff) {
        if (idx < l || idx > r)
            return;
        if(l == r) {
            seg[node] += diff;
            return;
        }
        int mid = (l + r) / 2;
        update(node * 2, l, mid, idx, diff);
        update(node * 2 + 1, mid + 1, r, idx, diff);
        seg[node] = f(seg[node*2], seg[node*2+1]);
    }
};
